tests/php/src/Internal/DependencyManagement/AbstractServiceProviderTests.php

```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * AbstractServiceProviderTests class file.
+ *
+ * @package Automattic\WooCommerce\Tests\Internal\DependencyManagement
+ */
+
+namespace Automattic\WooCommerce\Tests\Internal\DependencyManagement;
+
+use Automattic\WooCommerce\Internal\DependencyManagement\AbstractServiceProvider;
+use Automattic\WooCommerce\Internal\DependencyManagement\ExtendedContainer;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithConstructorArgumentWithoutTypeHint;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithDependencies;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithPrivateConstructor;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\DependencyClass;
+use League\Container\Definition\DefinitionInterface;
+
+/**
+ * Tests for AbstractServiceProvider.
+ */
+class AbstractServiceProviderTests extends \WC_Unit_Test_Case {
+
+	/**
+	 * The system under test.
+	 *
+	 * @var AbstractServiceProvider
+	 */
+	private $sut;
+
+	/**
+	 * The container used for tests.
+	 *
+	 * @var ExtendedContainer
+	 */
+	private $container;
+
+	/**
+	 * Runs before each test.
+	 */
+	public function setUp() {
+		$this->container = new ExtendedContainer();
+
+		$this->sut = new class() extends AbstractServiceProvider {
```

Comment:
Given that the underlying function is only "protected," do you think it might be better to use `ReflectionClass` and change the accessibility of the method? I just wonder if maybe anonymous classes might be too fragile?

Reply:
We are not only changing the visibility of a method here, we are also implementing an abstract method. Maybe we could also do that using reflection but I think that would be a mess.

Also if the anonymous class turns out to be problematic then we're going to notice pretty soon (failing tests) and then we can go and think about alternatives. 🙂